File: exercise/scala/day12/src/main/scala/gifts/Santa.scala
package gifts

import scala.collection.mutable.ListBuffer

case class Toy(description: String)

class Child(val name: String, val behavior: String) {
  private var wishlist: List[Toy] = List()

  def setWishlist(firstChoice: Toy, secondChoice: Toy, thirdChoice: Toy): Unit = {
    wishlist = List(firstChoice, secondChoice, thirdChoice)
  }

  def getWishlist: List[Toy] = wishlist
}

class Santa {
  private val childrenRepository = ListBuffer[Child]()

  def chooseToyForChild(childName: String): Option[Toy] = {
    val foundChild = childrenRepository.find(_.name == childName)

    if (foundChild.isEmpty) throw new NoSuchElementException("No such child found")

    val child = foundChild.get
    if (child.behavior == "naughty") {
      child.getWishlist.lastOption
    } else if (child.behavior == "nice") {
      child.getWishlist.lift(1)
    } else if (child.behavior == "very nice") {
      child.getWishlist.headOption
    } else {
      None
    }
  }

  def addChild(child: Child): Unit = {
    childrenRepository += child
  }
}

File: exercise/ts/day12/src/gifts/Santa.ts
import { Child } from './Child';
import { Toy } from './Toy';

export class Santa {
    private readonly childrenRepository: Child[] = [];

    addChild(child: Child): void {
        this.childrenRepository.push(child);
    }

    chooseToyForChild(childName: string): Toy | undefined {
        const foundChild = this.childrenRepository.find(child => child.name === childName);

        if (!foundChild) {
            throw new Error('No such child found');
        }

        if (foundChild.behavior === 'naughty') {
            return foundChild.wishlist[2];
        } else if (foundChild.behavior === 'nice') {
            return foundChild.wishlist[1];
        } else if (foundChild.behavior === 'very nice') {
            return foundChild.wishlist[0];
        }

        return undefined;
    }
}